File: lib/core/finance_math.dart
import 'dart:math' as math;

/// A single computed installment in an amortization schedule.
class AmortizationRow {
  final int installmentNo;
  final DateTime dueDate;
  final double emi;
  final double principal;
  final double interest;
  final double balance;

  const AmortizationRow({
    required this.installmentNo,
    required this.dueDate,
    required this.emi,
    required this.principal,
    required this.interest,
    required this.balance,
  });
}

class FinanceMath {
  /// Standard reducing-balance EMI.
  /// EMI = P * r * (1+r)^n / ((1+r)^n - 1), r = monthly rate.
  static double emi({
    required double principal,
    required double annualRatePct,
    required int tenureMonths,
  }) {
    if (tenureMonths <= 0) return 0;
    final r = annualRatePct / 12 / 100;
    if (r == 0) return principal / tenureMonths;
    final pow = math.pow(1 + r, tenureMonths);
    return principal * r * pow / (pow - 1);
  }

  /// Total interest paid over the full tenure.
  static double totalInterest({
    required double principal,
    required double annualRatePct,
    required int tenureMonths,
  }) {
    final e = emi(
      principal: principal,
      annualRatePct: annualRatePct,
      tenureMonths: tenureMonths,
    );
    return (e * tenureMonths) - principal;
  }

  /// Builds a full reducing-balance schedule with due dates anchored to [dueDay].
  static List<AmortizationRow> schedule({
    required double principal,
    required double annualRatePct,
    required int tenureMonths,
    required DateTime startDate,
    required int dueDay,
  }) {
    final rows = <AmortizationRow>[];
    final r = annualRatePct / 12 / 100;
    final e = emi(
      principal: principal,
      annualRatePct: annualRatePct,
      tenureMonths: tenureMonths,
    );
    var balance = principal;
    for (var i = 1; i <= tenureMonths; i++) {
      final interest = balance * r;
      var principalPart = e - interest;
      // Last installment: clear any rounding residue.
      if (i == tenureMonths || principalPart > balance) {
        principalPart = balance;
      }
      balance = math.max(0, balance - principalPart);
      rows.add(AmortizationRow(
        installmentNo: i,
        dueDate: _dueDateFor(startDate, i, dueDay),
        emi: i == tenureMonths ? principalPart + interest : e,
        principal: principalPart,
        interest: interest,
        balance: balance,
      ));
    }
    return rows;
  }

  static DateTime _dueDateFor(DateTime start, int monthOffset, int dueDay) {
    final base = DateTime(start.year, start.month + monthOffset, 1);
    final lastDay = DateTime(base.year, base.month + 1, 0).day;
    return DateTime(base.year, base.month, math.min(dueDay, lastDay));
  }

  /// Future value of a recurring SIP (annuity due not assumed; ordinary).
  static double sipFutureValue({
    required double monthly,
    required double annualRatePct,
    required int months,
  }) {
    final r = annualRatePct / 12 / 100;
    if (r == 0) return monthly * months;
    return monthly * ((math.pow(1 + r, months) - 1) / r) * (1 + r);
  }

  /// Maturity of a lump-sum at compound interest (annual compounding).
  static double compoundMaturity({
    required double principal,
    required double annualRatePct,
    required double years,
  }) {
    return principal * math.pow(1 + annualRatePct / 100, years);
  }
}

File: expense-manager-react/src/lib/finance-math.ts
import type { AmortizationRow } from '../types';
import { addMonths, setDate } from 'date-fns';

export function emi({
  principal,
  annualRatePct,
  tenureMonths,
}: {
  principal: number;
  annualRatePct: number;
  tenureMonths: number;
}): number {
  if (tenureMonths <= 0) return 0;
  if (annualRatePct <= 0) return principal / tenureMonths;

  const r = annualRatePct / 12 / 100;
  const n = tenureMonths;
  const factor = Math.pow(1 + r, n);
  const amount = principal * r * factor / (factor - 1);
  return Math.round(amount * 100) / 100;
}

export function totalInterest({
  principal,
  annualRatePct,
  tenureMonths,
}: {
  principal: number;
  annualRatePct: number;
  tenureMonths: number;
}): number {
  const monthlyEmi = emi({ principal, annualRatePct, tenureMonths });
  return monthlyEmi * tenureMonths - principal;
}

export function schedule({
  principal,
  annualRatePct,
  tenureMonths,
  startDate,
  dueDay = 5,
}: {
  principal: number;
  annualRatePct: number;
  tenureMonths: number;
  startDate: Date;
  dueDay?: number;
}): AmortizationRow[] {
  const rows: AmortizationRow[] = [];
  const monthlyEmi = emi({ principal, annualRatePct, tenureMonths });
  const r = annualRatePct / 12 / 100;
  let balance = principal;

  for (let i = 1; i <= tenureMonths; i++) {
    const interest = Math.round(balance * r * 100) / 100;
    const principalPart = Math.round((monthlyEmi - interest) * 100) / 100;
    balance = Math.round((balance - principalPart) * 100) / 100;

    const dueDate = addMonths(startDate, i);
    const finalDueDate = setDate(dueDate, Math.min(dueDay, 28));

    rows.push({
      installmentNo: i,
      dueDate: finalDueDate,
      emi: monthlyEmi,
      principal: principalPart,
      interest,
      balance: Math.max(0, balance),
    });
  }

  return rows;
}

export function sipFutureValue({
  monthly,
  annualRatePct,
  months,
}: {
  monthly: number;
  annualRatePct: number;
  months: number;
}): number {
  if (months <= 0) return 0;
  if (annualRatePct <= 0) return monthly * months;

  const r = annualRatePct / 12 / 100;
  const factor = Math.pow(1 + r, months);
  return Math.round(monthly * ((factor - 1) / r) * (1 + r) * 100) / 100;
}

export function compoundMaturity({
  principal,
  annualRatePct,
  years,
}: {
  principal: number;
  annualRatePct: number;
  years: number;
}): number {
  if (years <= 0) return principal;
  if (annualRatePct <= 0) return principal;

  const r = annualRatePct / 100;
  return Math.round(principal * Math.pow(1 + r, years) * 100) / 100;
}
